acme/authority.go

```
@@ -254,32 +256,209 @@ func (a *Authority) GetAuthz(p provisioner.Interface, accID, authzID string) (*A
+		return nil, Wrap(err, "error saving challenge")
+	}
+	ch = up
+
+	v, err := a.validate(ch, jwk)
+	// An error here is non-recoverable. Recoverable errors are set on the challenge object
+	// and should not be returned directly.
+	if err != nil {
+		return nil, Wrap(err, "error attempting challenge validation")
+	}
+	err = v.save(a.db, ch)
+	if err != nil {
+		return nil, Wrap(err, "error saving challenge")
+	}
+	ch = v
+
+	switch ch.getStatus() {
+	case StatusValid, StatusInvalid:
+		break
+	case StatusProcessing:
+		if ch.getRetry().Active() {
+			time.AfterFunc(retryInterval, func() {
+				a.RetryChallenge(ch.getID())
+			})
+		}
+	default:
+		e := errors.Errorf("post-validation challenge in unexpected state, %s", ch.getStatus())
+		return nil, ServerInternalErr(e)
+	}
+	return ch.toACME(a.dir, p)
+}
+
+// The challenge validation process is specific to the type of challenge (dns-01, http-01, tls-alpn-01).
+// But, we still pass generic "options" to the polymorphic validate call.
+func (a *Authority) validate(ch challenge, jwk *jose.JSONWebKey) (challenge, error) {
 	client := http.Client{
 		Timeout: time.Duration(30 * time.Second),
 	}
 	dialer := &net.Dialer{
 		Timeout: 30 * time.Second,
 	}
-	ch, err = ch.validate(a.db, jwk, validateOptions{
+	return ch.clone().morph().validate(jwk, validateOptions{
 		httpGet:   client.Get,
 		lookupTxt: net.LookupTXT,
 		tlsDial: func(network, addr string, config *tls.Config) (*tls.Conn, error) {
 			return tls.DialWithDialer(dialer, network, addr, config)
 		},
 	})
+}
+
+const retryInterval = 12 * time.Second
+
+// RetryChallenge behaves similar to ValidateChallenge, but simply attempts to perform a validation and
+// write update the challenge record in the db if the challenge has remaining retry attempts.
+//
+// see: ValidateChallenge
+func (a *Authority) RetryChallenge(chID string) {
+	ch, err := getChallenge(a.db, chID)
 	if err != nil {
-		return nil, Wrap(err, "error attempting challenge validation")
+		return
```

Let's at least log the error here, otherwise we're failing silently, right?